tls/extensions/s2n_client_key_share.c

```diff
@@ -98,6 +95,62 @@ static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s
     return S2N_SUCCESS;
 }
 
+static int s2n_send_hrr_keyshare(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+    const struct s2n_ecc_named_curve *server_negotiated_curve = NULL;
+    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
+
+    const struct s2n_ecc_preferences *ecc_pref = NULL;
+    GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
+    notnull_check(ecc_pref);
+
+    server_negotiated_curve = conn->secure.server_ecc_evp_params.negotiated_curve;
+    ENSURE_POSIX(server_negotiated_curve != NULL, S2N_ERR_BAD_KEY_SHARE);
+
+    bool has_supported_curve = false;
+    size_t supported_curve_index = 0;
+
+    for (size_t i = 0; i < ecc_pref->count; i++) {
+        if (ecc_pref->ecc_curves[i]->iana_id == server_negotiated_curve->iana_id) {
+            supported_curve_index = i;
+            has_supported_curve = true;
```

Comment:
Would it be cleaner if you just generated the keyshare inside this for loop? If this is the server_negotiated_curve, generate the keyshare, otherwise free the keyshare?

Reply:
I think it will be confusing to read. If the ClienHello1 failed, our keyshares were not successful and we should first clear the old list of keyshares prior to populating a new list of keyshares to be sent.

By generation the keyshares inside this loop we are not gaining any performance either as we break as soon as we find a server_negotiated_curve.